server/lib/nl/fulfillment/fulfiller.py

```diff
@@ -294,3 +298,49 @@ def _update_chart_vars_for_rig(state: PopulateState):
                              list(added_svs))
 
   state.chart_vars_map = updated_chart_vars_map
+
+
+#
+# Update chart vars map for toolformer rag mode by changing the order of chart
+# vars. Promote svs immediately after a topic to before the topic if they are
+# found in the topic.
+#
+def _update_chart_vars_for_rag(state: PopulateState):
+  # ordered list of chart vars map items
+  ordered_cv_map = list(state.chart_vars_map.items())
+
+  # go over the chart_vars_map and update ordering
+  updated_chart_vars_map = OrderedDict()
+  promoted_svs = set()
+  for idx, (var, cv_list) in enumerate(ordered_cv_map):
+    # handle non topic vars
+    if not var.startswith(_TOPIC_PREFIX):
+      # if var hasn't already been added to the updated chart vars map, add it
+      if not var in updated_chart_vars_map:
+        updated_chart_vars_map[var] = cv_list
+      continue
+
+    # get all the svs in the topic
+    topic_svs = set()
+    for cv in cv_list:
+      topic_svs.update(cv.svs)
+
+    # add immediate next svs that are part of the current topic so that they are
+    # promoted to before the current topic.
+    for next_var_idx in range(idx + 1, len(ordered_cv_map)):
+      next_var, next_var_cv_list = ordered_cv_map[next_var_idx]
+      # break if not a sv in the topic because we only want to add the immediate
+      # next svs
+      if next_var.startswith(_TOPIC_PREFIX) or not next_var in topic_svs:
```

Comment:
is the first check redundant?

Reply:
I wasn't sure if chartvar.svs can include topic vars .. if not, I can remove the first check

Reply:
I think chart_vars `svs` are variables (ensured in lib/nl/explore/topic.py), but fine to keep.